### PR TITLE
Fixing docker caching.

### PR DESCRIPTION
**Pull request by dessalines**
Using your old docker caching method, it works better. Still not as ideal as having a `cargo build --cache-deps`

### CODE REVIEW

File: docker/dev/Dockerfile [hunk 1]
```diff
@@ -17,14 +17,19 @@ WORKDIR /app
...
-   && echo 'fn main() { println!("Dummy") }' > ./src/bin/main.rs
+   && echo 'fn main() { println!("Dummy")  }' > ./src/bin/main.rs \
+   && cp ./src/bin/main.rs ./lemmy_db/src/main.rs \
+   && cp ./src/bin/main.rs ./lemmy_utils/src/main.rs
 RUN cargo build
-RUN find target/debug -type f -name "$(echo "lemmy_server" | tr '-' '_')*" -exec touch -t 200001010000 {} +
+RUN rm -f ./target/x86_64-unknown-linux-musl/debug/deps/lemmy_server*
 COPY server/src ./src/
+COPY server/lemmy_db ./lemmy_db/
+COPY server/lemmy_utils ./lemmy_utils/
```
Comment (Nutomic) [inline docker/dev/Dockerfile:32]: Are these good for anything?

**Comment by Nutomic** [comment]
This isnt working properly for me. I compiled, then added a simple println() in main.rs and compiled again. But it recompiled all the dependencies of lemmy_server. Only dependencies of lemmy_utils and lemmy_db were cached. 

The same thing is happening if I change anything in lemmy_utils or lemmy_db. So it looks like dependencies of lemmy_utils and lemmy_db are cached fine, but direct dependencies of lemmy_server itself are never cached.

File: docker/dev/Dockerfile (hunk 1 above)
Comment (dessalines) [reply docker/dev/Dockerfile:32]: Yeah, the src folders aren't copied before, only the Cargo.toml s .

Comment (Nutomic) [reply docker/dev/Dockerfile:32]: Ah right, misunderstood this. Maybe put a comment what its doing?

**Comment by dessalines** [comment]
I haven't tried any changes to the main, but the other one was really annoying because any change in `lemmy_db` or `lemmy_utils` would re-cache everything. I'll see if I can find a better solution, because builds right now are 30+ minutes for me from scratch.

**Comment by Nutomic** [comment]
Okay if this is an improvement over what we have now, then we can merge it. Sucks that it is so hard to handle properly.

**Comment by dessalines** [comment]
A few more solutions got added here, I'm gonna try them out today.

https://stackoverflow.com/questions/54952867/cache-cargo-dependencies-in-a-docker-volume

**Comment by dessalines** [comment]
Okay I got it working finally.

File: docker/dev/Dockerfile [hunk 2]
```diff
@@ -10,27 +10,36 @@ RUN yarn install --pure-lockfile
 COPY ui /app/ui
 RUN yarn build
 
-FROM ekidd/rust-musl-builder:1.42.0-openssl11 as rust
+FROM ekidd/rust-musl-builder:nightly-2020-05-07 as rust
```
Comment (dessalines) [inline docker/dev/Dockerfile:13]: It needs nightly, because it uses some unstable features to build the dependency tree.

File: server/lemmy_db/Cargo.toml [hunk 3]
```diff
@@ -3,6 +3,10 @@ name = "lemmy_db"
 version = "0.1.0"
 edition = "2018"
 
+[lib]
+name = "lemmy_db"
+path = "src/lib.rs"
```
Comment (dessalines) [inline server/lemmy_db/Cargo.toml:8]: These were necessary too.

File: docker/dev/docker-compose.yml [hunk 4]
```diff
@@ -3,9 +3,7 @@ version: '3.3'
 services:
 
   lemmy:
-    build: 
-      context: ../../
-      dockerfile: docker/dev/Dockerfile
+    image: lemmy-dev:latest
```
Comment (Nutomic) [inline docker/dev/docker-compose.yml:6]: Why did you change this?

File: docker/dev/Dockerfile (hunk 2 above)
Comment (Nutomic) [reply docker/dev/Dockerfile:13]: This isnt ideal, our CI should ensure that Lemmy builds fine on latest stable and doesnt use any nightly only features. Maybe we need to add a separate Dockerfile for CI?
